score.go

```diff
@@ -15,31 +15,252 @@
-// debugScore is true, it also adds a debug string to the FileMatch. If raw is
-// -1, it is ignored. Otherwise, it is added to the debug string.
-func (m *FileMatch) addScore(what string, computed float64, raw float64, debugScore bool) {
-	if computed != 0 && debugScore {
-		var b strings.Builder
-		fmt.Fprintf(&b, "%s", what)
-		if raw != -1 {
-			fmt.Fprintf(&b, "(%s)", strconv.FormatFloat(raw, 'f', -1, 64))
+type chunkScore struct {
+	score      float64
+	debugScore string
+	bestLine   int
+}
+
+// scoreChunk calculates the score for each line in the chunk based on its candidate matches, and returns the score of
+// the best-scoring line, along with its line number.
+// Invariant: there should be at least one input candidate, len(ms) > 0.
+func (p *contentProvider) scoreChunk(ms []*candidateMatch, language string, opts *SearchOptions) (chunkScore, []*Symbol) {
+	nl := p.newlines()
+
+	var bestScore lineScore
+	bestLine := 0
+	var symbolInfo []*Symbol
+
+	start := 0
+	currentLine := -1
+	for i, m := range ms {
+		lineNumber := -1
+		if !m.fileName {
+			lineNumber = nl.atOffset(m.byteOffset)
+		}
+
+		// If this match represents a new line, then score the previous line and update 'start'.
+		if i != 0 && lineNumber != currentLine {
+			score, si := p.scoreLine(ms[start:i], language, currentLine, opts)
+			symbolInfo = append(symbolInfo, si...)
+			if score.score > bestScore.score {
+				bestScore = score
+				bestLine = currentLine
+			}
+			start = i
+		}
+		currentLine = lineNumber
+	}
+
+	// Make sure to score the last line
+	line, si := p.scoreLine(ms[start:], language, currentLine, opts)
+	symbolInfo = append(symbolInfo, si...)
+	if line.score > bestScore.score {
+		bestScore = line
+		bestLine = currentLine
+	}
+
+	cs := chunkScore{
+		score:    bestScore.score,
+		bestLine: bestLine,
+	}
+	if opts.DebugScore {
+		cs.debugScore = fmt.Sprintf("%s, (line: %d)", bestScore.debugScore, bestLine)
+	}
+	return cs, symbolInfo
+}
+
+type lineScore struct {
+	score      float64
+	debugScore string
+}
+
+// scoreLine calculates a score for the line based on its candidate matches.
+// Invariants:
+// - All candidate matches are assumed to come from the same line in the content.
+// - If this line represents a filename, then lineNumber must be -1.
+// - There should be at least one input candidate, len(ms) > 0.
+func (p *contentProvider) scoreLine(ms []*candidateMatch, language string, lineNumber int, opts *SearchOptions) (lineScore, []*Symbol) {
+	if opts.UseBM25Scoring {
+		score, symbolInfo := p.scoreLineBM25(ms, lineNumber)
+		ls := lineScore{score: score}
+		if opts.DebugScore {
+			ls.debugScore = fmt.Sprintf("tfScore:%.2f, ", score)
+		}
+		return ls, symbolInfo
+	}
+
+	score := 0.0
+	what := ""
+	addScore := func(w string, s float64) {
+		if s != 0 && opts.DebugScore {
+			what += fmt.Sprintf("%s:%.2f, ", w, s)
+		}
+		score += s
+	}
+
+	filename := p.data(true)
+	var symbolInfo []*Symbol
+
+	var bestLine lineScore
```

Comment:
nit: `bestLine` isn't the best name if we're already operating on a single line - `bestScore` might be better.